### PR TITLE
`endpoint` proc-macro

### PR DESCRIPTION
**Pull request by YaronWittenstein**
# Motivation

Adding `[endpoint]` proc-macro attribute. It should ease the implementation of apps' endpoints.

This macro implements:

* Auto-parsing of the input calldata to Rust native types (see PR: #181)

* Will generate in a future PR the API schema in a JSON format (will be consumed by `smapp`). For now, it will gather the required data in compile-time.

### CODE REVIEW

File: crates/svm-sdk/macros/tests/endpoint.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        fn test<F>(f: F) where F: FnOnce() {
+            // run the tests in a serial manner
+            let guard = TEST_LOCK.lock().unwrap();
+
+            let host = MockHost::instance();
+            host.reset();
+
+            f();
+
+            drop(guard)
```
Comment (moshababo) [inline crates/svm-sdk/macros/tests/endpoint.rs:46]: Isn't `guard` going to be dropped here anyway?

File: crates/svm-sdk/macros/tests/storage.rs [hunk 2]
```diff
@@ -0,0 +1,271 @@
+// use cfg_if::cfg_if;
+
```
Comment (moshababo) [inline crates/svm-sdk/macros/tests/storage.rs:2]: Why commented?

File: crates/svm-sdk/storage/src/mock.rs [hunk 3]
```diff
@@ -173,7 +173,7 @@ fn storage() -> MutexGuard<'static, InnerStorage> {
 mod tests {
     use super::*;
 
-    use crate::memory::alloc;
+    use svm_sdk_alloc::alloc;
```
Comment (moshababo) [inline crates/svm-sdk/storage/src/mock.rs:176]: `STORAGE` instance init/access pattern is different than `HOST` - If there's no reason for that, it should be unified.

File: crates/svm-sdk/alloc/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+//! Using this crate when writing SVM Templates in Rust isn't mandatory but should be very useful.
+//!
+//! The crate is compiled with `![no_std]` (no Rust stdlib) annotation in order to reduce the compiled WASM size.
+
+#![allow(missing_docs)]
+#![allow(unused)]
+#![allow(dead_code)]
+#![allow(unreachable_code)]
+
+extern crate alloc;
```
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-sdk/alloc/src/lib.rs:14]: I prefer to keep the current name for now.
I don't expect it to change in the future.

**Review by moshababo — APPROVED** [review]

**Pull request merged**